Add type hints

larsimmisch opened the pull request:
This should address #58

ossilator commented:
what was wrong with the previous PR?

larsimmisch commented:
I was just too tired to rebase it to the new `main`, so I created a new branch/PR. 🤷‍♂️

ossilator commented:
i'm confused. you needed to rebase locally anyway, and updating the PR ist just a forced push into the existing branch.

ossilator commented:
some random observations from a quick scan:
- you should probably rebase and cover the two new functions i added docu for
- you're being inconsistent about the whitespace after colons
- the trailing whitespace removal should be committed separately. given that you have set up your editor to do that, it's probably best if you just touch all files once, so it's out of the way.

how are the hints rendered in html? do the prototypes look just as you provide them, or does it do something clever? i guess the former, as you need to provide the info redundantly ...

RonaldAJ commented on doc/libalsaaudio.rst:
```diff
@@ -267,37 +267,29 @@ PCM objects have the following methods:
-   otherwise they represent those stored by pyalsaaudio, and their keys
-   are prefixed with ' (call value) '.
-
    *New in 0.9.1*
```
*New in 0.9.1* belonged to PCM.info so should be removed

RonaldAJ replied:
Can't comment on the technical changes. I would have to read up on type hints before I can.

ossilator commented on alsaaudio.pyi:
```diff
@@ -0,0 +1,128 @@
+def mixers(cardindex:int=-1, device:str='default') -> list[str]: ...
+def asoundlib_version() -> str: ...
+
+class PCM:
```
some functions are definitely still missing here.

larsimmisch replied:
I didn't document the ones that return `None` and have no arguments, is that it? 

I think type hints are great, but as they're optional, less might be more. In any case, I'll try to double check for consistency

ossilator replied:
i didn't check, i assume you're right.
but the argument doesn't convince me. firstly, that inconsistency is confusing - the question whether this was omitted by accident will come up each time someone notices it. secondly, i think that file may be used for auto-completion by some IDEs, so having complete declarations is helpful regardless.

ossilator commented on doc/libalsaaudio.rst:
```diff
@@ -410,22 +410,15 @@ PCM objects have the following methods:
 
    *New in 0.10*
 
-.. method:: PCM.close()
```
if this is legit (is it?), then it definitely belongs into a separate commit.

larsimmisch replied:
It's legit. `close` wasn't deleted, but moved to the end. Creating a separate commit for that seems overkill to me, though.

ossilator replied:
> It's legit. close wasn't deleted, but moved to the end.

yes, i've seen that later.

> Creating a separate commit for that in retrospect seems overkill to me, though.

but this is exactly what good commit hygiene is about. it would have avoided me asking that "stupid" question.

and fixing it would take about 30 seconds in git gui (plus however much you need to type the commit message).

ossilator commented on doc/conf.py:
```diff
@@ -35,7 +35,6 @@
 # extensions coming with Sphinx (named 'sphinx.ext.*') or your custom
 # ones.
 extensions = []
-
```
how did that get past the whitespace cleanup?

ossilator reviewed (changes requested):
the commit message of the second commit is sort of backwards - you're adding type hints to the docu, not the other way around.
but i'd actually squash the first two commits.

the last commit could use a description in the commit message. one sentence should be enough.

note that the comment about whitespace inconsistency still applies.

larsimmisch commented:
> but i'd actually squash the first two commits.

Yeah, clearer, done.

ossilator commented:
totally predictable comment from me: the .pyi change should be squashed onto the first commit.

as for the .c change, i'd be in favor of just dropping it - it's churn that makes following the history (blame) a lot harder, for very little benefit, imo.

if you insist on keeping it, then you should factor it out into a commit that goes at the beginning of the series, so it's in front of the addition that replicates the same order.

what does the "fix todo" refer to?

larsimmisch commented:
Hi @ossilator,

I've addressed your feedback (thanks) and did another force-push.

What I haven't done yet is validating this change, i.e. *see* the type hints in a native Linux development environment. Maybe I'll install a recent python-mode for Emacs.

> totally predictable comment from me: the .pyi change should be squashed onto the first commit.
:-)

> as for the .c change, i'd be in favor of just dropping it - it's churn that makes following the history (blame) a lot harder, for very little benefit, imo.

I get your point, but I'd rather have `getchannels/setchannels` together.

> if you insist on keeping it, then you should factor it out into a commit that goes at the beginning of the series, so it's in front of the addition that replicates the same order.

Done.

ossilator commented:
structure-wise it appears correct now.

you're still a bit inconsistent with the whitespace in the type hints. try to standardize on one style, preferably no space before the colon and one space after the colon. though there is a challenge with default params then, because "foo: int=5" looks kinda stupid. maybe surround the equal sign with spaces as well? did the python community develop a canonical style for this?

larsimmisch commented:
@ossilator I've made the whitespace consistent like you suggested (thanks) and squashed the change into the second commit.

Maybe more importantly, I verified that the type hints are shown using VSCode/Pylance. I noticed that IDE support isn't great without the docstrings, because parameters aren't shown. I'll try to experiment with Sphinx autodoc, but that would be another PR.

> though there is a challenge with default params then, because "foo: int=5" looks kinda stupid. maybe surround the equal sign with spaces as well? did the python community develop a canonical style for this?

To my eyes `foo: int=5` looks ok. Also, PEP 8 wants no spaces around the equal sign for default parameters, so I used none here.

ossilator commented on doc/libalsaaudio.rst:
```diff
@@ -10,7 +10,7 @@
 The :mod:`alsaaudio` module defines functions and classes for using ALSA.
 
-.. function:: pcms(pcmtype=PCM_PLAYBACK)
+.. function:: pcms(pcmtype:int=PCM_PLAYBACK) ->list[str]
```
...

ossilator commented on alsaaudio.pyi:
```diff
@@ -0,0 +1,135 @@
+VOLUME_UNITS_RAW: int
+VOLUME_UNITS_DB: int
+
+def pcms(pcmtype:int) -> list[str]: ...
```
missing space

ossilator commented on alsaaudio.pyi:
```diff
@@ -0,0 +1,135 @@
+	def setformat(format: int) -> int: ...
+	def setperiodsize(period: int) -> int: ...
+	def read() -> tuple[int, bytes]: ...
+	def write(data:bytes) -> int: ...
```
missing space

ossilator commented on alsaaudio.pyi:
```diff
@@ -0,0 +1,135 @@
+	def read() -> tuple[int, bytes]: ...
+	def write(data:bytes) -> int: ...
+	def avail() -> int: ...
+	def pause(enable:bool=True) -> int: ...
```
...

ossilator commented on doc/libalsaaudio.rst:
```diff
@@ -625,13 +625,13 @@ Mixer objects have the following methods:
    control.
 
-.. method:: Mixer.setenum(index)
+.. method:: Mixer.setenum(index:int) -> None
```
...

ossilator reviewed (commented):
you seem to still have missed some in the .pyi file. but did you touch the .rst file at all? there are still plenty; i just stopped marking them at some point.
you can easily locate them by regex-grepping for `:[^ ]`.

note also the still open request about close().

larsimmisch commented:
Hi @ossilator,

I went through the PR again and fixed all whitespace issues. And a few missing type hints. I also undid the move of `close` (see https://github.com/larsimmisch/pyalsaaudio/pull/143#discussion_r1496436396). Hope this is complete/consistent now.

ossilator commented on doc/libalsaaudio.rst:
```diff
@@ -675,14 +677,14 @@ Mixer objects have the following methods:
    This method will fail if the mixer has no playback switch capabilities.
 
-.. method:: Mixer.setmute(mute, [channel])
+.. method:: Mixer.setmute(mute: bool, channel: int|None=None) -> None
```
missing spaces around `|`.

ossilator commented on doc/libalsaaudio.rst:
```diff
@@ -692,14 +694,14 @@ Mixer objects have the following methods:
    This method will fail if the mixer has no capture switch capabilities.
 
-.. method:: Mixer.setrec(capture, [channel])
+.. method:: Mixer.setrec(capture: int, channel: int|None=None) -> None
```
missing spaces around `|`.

ossilator reviewed (changes requested):
sorry, still a few issues left. :-(

given how easy i find it to spot these issues, i wonder why you don't. maybe you have an inappropriate font or font size configured? or is it just a perception difference? i find these breaks in the pattern jarring, so i just can't overlook them.

larsimmisch commented on doc/libalsaaudio.rst:
```diff
@@ -658,7 +660,7 @@ Mixer objects have the following methods:
    :const:`VOLUME_UNITS_RAW`, or :const:`VOLUME_UNITS_DB`.
 
-.. method:: Mixer.setvolume(volume, channel=None, pcmtype=PCM_PLAYBACK, units=VOLUME_UNITS_PERCENTAGE)
+.. method:: Mixer.setvolume(volume: int, pcmtype: int=PCM_PLAYBACK, units: int=VOLUME_UNITS_PERCENTAGE, channel: int|None) -> None
```
That was deliberate (except for one inconsistency).

But I now see that https://docs.python.org/3/library/typing.html mostly has spaces before and after `|` (that document also uses `|` without surrounding spaces, but just in one case).

This reads ok to me with the exception of, say, `channel: int | None = None`. In these cases, I'll use `channel: (int | None) = None`.

ossilator replied:
is that legal? i didn't read the document, but a search for `|` revealed no precedent for using parentheses.

larsimmisch replied:
PyLance does not complain, but does complain when obvious syntax errors are present (like mismatched brackets).

ossilator replied:
a more convincing test is just running it through real python. seems to work.

larsimmisch commented on doc/libalsaaudio.rst:
```diff
@@ -599,7 +601,7 @@ Mixer objects have the following methods:
    ========================  ================
 
-.. method:: Mixer.getenum()
+.. method:: Mixer.getenum() -> tuple[ str, list[str]]
```
Fixed, thanks.

ossilator commented on doc/libalsaaudio.rst:
```diff
@@ -523,7 +525,7 @@ Mixer Objects
 Mixer objects provides access to the ALSA mixer API.
 
-.. class:: Mixer(control='Master', id=0, cardindex=-1, device='default')
+.. class:: Mixer(control: str = 'Master', id: int = 0, cardindex: int = -1, device str = 'default') -> Mixer
```
missing colon after 'device'.

larsimmisch commented:
@ossilator thanks for the reviews
